pkg/virtctl/vmexport/vmexport.go

```diff
@@ -0,0 +1,639 @@
+	req.Header.Set(httpHeader, token)
+
+	return httpClient.Do(req)
+}
+
+// getHTTPClient assigns the default, non-mocked HTTP client
+func getHTTPClient(transport *http.Transport) *http.Client {
+	if transport == nil {
+		transport = &http.Transport{
+			TLSClientConfig: &tls.Config{
+				InsecureSkipVerify: true,
+			},
+		}
+	}
+
+	client := &http.Client{Transport: transport}
+	return client
+}
+
+// copyFileWithProgressBar serves as a wrapper to copy the file with a progress bar
+func copyFileWithProgressBar(output *os.File, resp *http.Response) error {
+	// TODO: Unable to get file size when downloading using the compressed URL
+	filesize, _ := strconv.Atoi(resp.Header.Get("Content-Length"))
+
+	// Initiate progress bar
+	bar := pb.New(filesize).SetUnits(pb.U_BYTES)
+	rd := bar.NewProxyReader(resp.Body)
+	fmt.Println("Downloading file...")
+	bar.Start()
+
+	_, err := io.Copy(output, rd)
+	bar.Finish()
+	return err
+}
+
+// getOrCreateTokenSecret obtains a token secret to be used along with the virtualMachineExport
+func getOrCreateTokenSecret(client kubecli.KubevirtClient, namespace string) (*k8sv1.Secret, error) {
+	// Generate a random, 20 char string to be used as a token
+	token := rand.String(20)
+
+	secret := &k8sv1.Secret{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      getExportSecretName(),
+			Namespace: namespace,
+		},
+		Type: k8sv1.SecretTypeOpaque,
+		Data: map[string][]byte{
+			secretTokenEntry: []byte(token),
+		},
+	}
+
+	_, err := client.CoreV1().Secrets(namespace).Create(context.Background(), secret, metav1.CreateOptions{})
+	if err != nil && !k8serrors.IsAlreadyExists(err) {
+		return nil, err
+	}
+
+	return secret, nil
+}
+
+// deleteExportSecret deletes the secret assigned to the virtualMachineExport
+func deleteExportSecret(client kubecli.KubevirtClient, namespace string) error {
+	secretName := getExportSecretName()
+	if err := client.CoreV1().Secrets(namespace).Delete(context.TODO(), secretName, metav1.DeleteOptions{}); err != nil {
+		if !k8serrors.IsNotFound(err) {
+			return err
+		}
+		return nil
+	}
+
+	fmt.Printf("Secret '%s/%s' deleted succesfully\n", namespace, secretName)
+	return nil
+}
+
+// getTokenFromSecret extracts the token from the secret specified on the virtualMachineExport
+func getTokenFromSecret(client kubecli.KubevirtClient, namespace string) (string, error) {
+	secretName := getExportSecretName()
+	secret, err := client.CoreV1().Secrets(namespace).Get(context.Background(), secretName, metav1.GetOptions{})
+	if err != nil {
+		return "", err
+	}
+
+	token := secret.Data[secretTokenEntry]
+	return string(token), nil
+}
+
+// handleCreateFlags ensures that only compatible flag combinations are used with 'create' or with 'download' with the --create flag
+func handleCreateFlags(args []string) error {
+	if vm == "" && snapshot == "" && pvc == "" {
+		return fmt.Errorf(ErrRequiredExportType)
+	}
+
+	if vm != "" {
+		resourceKind = "VirtualMachine"
```

Comment:
As I mentioned above, the API group is different for the different source types, so it might make sense to use a TypedLocalObjectReference instead of putting the individual fields in a string. Also maybe return the object instead of storing it in a global variable.